core/consensus/babe/src/lib.rs

```diff
@@ -14,58 +14,140 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
 
+//! # BABE consensus
+//!
 //! BABE (Blind Assignment for Blockchain Extension) consensus in substrate.
-#![forbid(warnings, unsafe_code, missing_docs)]
-
+//!
+//! # Stability
+//!
+//! This crate is highly unstable and experimental.  Breaking changes may
+//! happen at any point.  This crate is also missing features, such as banning
+//! of malicious validators, that are essential for a production network.
+#![forbid(unsafe_code, missing_docs)]
+#![deny(warnings)]
+extern crate core;
 pub use babe_primitives::*;
+use runtime_primitives::{generic, generic::BlockId, Justification};
+use runtime_primitives::traits::{
+	Block, Header, Digest, DigestItemFor, DigestItem, ProvideRuntimeApi, AuthorityIdFor,
+};
+use std::{sync::Arc, u64, fmt::Debug};
 use parity_codec::{Decode, Encode, Input};
-use runtime_primitives::generic;
-use primitives::sr25519::{
-	Public,
-	Signature,
-	LocalizedSignature,
+use primitives::{
+	crypto::Pair,
+	sr25519::{Public, Signature, LocalizedSignature, self},
 };
-
+use merlin::Transcript;
+use inherents::{InherentDataProviders, InherentData, RuntimeString};
+use substrate_telemetry::{telemetry, CONSENSUS_TRACE, CONSENSUS_DEBUG, CONSENSUS_WARN, CONSENSUS_INFO};
 use schnorrkel::{
-	SecretKey as Secret,
-	vrf::{VRFProof, VRF_PROOF_LENGTH},
+	keys::Keypair,
+	vrf::{
+		VRFProof, VRFProofBatchable, VRFInOut, VRFOutput,
+		VRF_OUTPUT_LENGTH, VRF_PROOF_LENGTH,
+	},
 	PUBLIC_KEY_LENGTH, SIGNATURE_LENGTH,
 };
+pub use consensus_common::SyncOracle;
+use authorities::AuthoritiesApi;
+use consensus_common::{self, Authorities, BlockImport, Environment, Proposer,
+	ForkChoiceStrategy, ImportBlock, BlockOrigin, Error as ConsensusError,
+};
+use srml_babe::{
+	BabeInherentData,
+	timestamp::{TimestampInherentData, InherentType as TimestampInherent}
+};
+use consensus_common::well_known_cache_keys;
+use consensus_common::import_queue::{Verifier, BasicQueue};
+use client::{
+	ChainHead,
+	block_builder::api::BlockBuilder as BlockBuilderApi,
+	blockchain::ProvideCache,
+	runtime_api::ApiExt,
+	error::Result as CResult,
+	backend::AuxStore,
+};
+use slots::CheckedHeader;
+use futures::{Future, IntoFuture, future};
+use tokio::timer::Timeout;
+use log::{error, warn, debug, info, trace};
+
+use slots::{SlotWorker, SlotInfo, SlotCompatible, slot_now};
 
 /// A BABE seal.  It includes:
-/// 
+///
 /// * The public key
 /// * The VRF proof
 /// * The signature
 /// * The slot number
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct BabeSeal {
+	vrf_output: VRFOutput,
 	proof: VRFProof,
 	signature: LocalizedSignature,
 	slot_num: u64,
 }
 
+/// The prefix used by BABE for its VRF keys.
+pub const BABE_VRF_PREFIX: &'static [u8] = b"substrate-babe-vrf";
+
+macro_rules! babe_assert_eq {
+	($a: expr, $b: expr) => {
+		({
```

Comment:
are these parentheses redundant?

Reply:
Not sure.  I included them just in case.